### PR TITLE
New package pocl

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

**Comment by eschnett** [comment]
Note that this pull request contains necessary modifications to two other packages as well. I also just submitted separate pull requests for these.

File: var/spack/repos/builtin/packages/pocl/package.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    depends_on("llvm @3.7:3.9", when="@0.14")
+    depends_on("llvm @3.7:3.8", when="@0.13")
+    depends_on("llvm @3.2:3.7", when="@0.12")
+    depends_on("llvm @3.2:3.6", when="@0.11")
+    depends_on("llvm @3.2:3.5", when="@0.10")
+
+    variant("distro", default=False,
+            description=("Support several CPU architectures, " +
+                         "suitable e.g. in a build that will be made " +
+                         "available for download"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pocl/package.py]: This doesn't really need to be split into multiple lines if you don't want to. Lines that start with `variant` are [excluded](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/cmd/flake8.py#L60) from the line length checks.

File: var/spack/repos/builtin/packages/pocl/package.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            args += ["-DSTATIC_LLVM"]
+        if "+distro" in spec:
+            args += ["-DKERNELLIB_HOST_CPU_VARIANTS=distro"]
+        args += ["-DENABLE_ICD=%s" % ("ON" if "+icd" in spec else "OFF")]
+        return args
+
+    def install(self, spec, prefix):
+        with working_dir(self.build_directory):
+            make(*self.install_targets)
+            os.symlink("OpenCL", join_path(self.prefix.include, "CL"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pocl/package.py]: It might be easier to keep the default `install` and add a:
```python
@run_after('install')
def symlink_opencl(self):
    with working_dir(self.build_directory):
        os.symlink("OpenCL", join_path(self.prefix.include, "CL"))
```
You can add multiple `@run_after('install')` methods if you want to.

File: var/spack/repos/builtin/packages/bison/package.py [hunk 3]
```diff
@@ -35,7 +35,8 @@ class Bison(AutotoolsPackage):
 
     version('3.0.4', 'a586e11cd4aff49c3ff6d3b6a4c9ccf8')
 
-    depends_on('m4', type='build')
+    # This is a run-time dependency as well
+    depends_on('m4')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/bison/package.py]: I would leave these changes out of this PR if you're going to submit them in another PR. Otherwise both PRs will be held up until we decide on what to do about this dependency, and the discussion will be split in multiple places.

File: var/spack/repos/builtin/packages/pocl/package.py (hunk 1 above)
Comment (eschnett) [inline var/spack/repos/builtin/packages/pocl/package.py]: Ok.

File: var/spack/repos/builtin/packages/pocl/package.py (hunk 2 above)
Comment (eschnett) [inline var/spack/repos/builtin/packages/pocl/package.py]: I assume all the `@run_after` are run in the order they are declared in the source code?

File: var/spack/repos/builtin/packages/bison/package.py (hunk 3 above)
Comment (eschnett) [inline var/spack/repos/builtin/packages/bison/package.py]: Removed.

File: var/spack/repos/builtin/packages/pocl/package.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    # These are the supported LLVM versions
+    depends_on("llvm @3.7:3.9", when="@master")
+    depends_on("llvm @3.7:3.9", when="@0.14")
+    depends_on("llvm @3.7:3.8", when="@0.13")
+    depends_on("llvm @3.2:3.7", when="@0.12")
+    depends_on("llvm @3.2:3.6", when="@0.11")
+    depends_on("llvm @3.2:3.5", when="@0.10")
+
+    variant("distro", default=False,
+            description="Support several CPU architectures, suitable e.g. in a build that will be made available for download")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pocl/package.py]: You'll have to either put this on the same line as `variant` or add `# noqa` to the end so flake8 skips it.

File: var/spack/repos/builtin/packages/pocl/package.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    version("0.10", "0096be4f595c7b5cbfa42430c8b3af6a")
+
+    # This is Github's pocl/pocl#373
+    patch("uint.patch", when="@:0.13")
+    patch("vecmathlib.patch", when="@:0.13")
+
+    depends_on("cmake @2.8.12:", type="build")
+    depends_on("hwloc")
+    # Yes, libtool is a run-time dependency
+    depends_on("libtool")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pocl/package.py]: Was this intended to be a run type dependency?

**Comment by alalazo** [comment]
Apparently this PR broke `develop`, still need to investigate why.

**Comment by alalazo** [comment]
[This](https://travis-ci.org/LLNL/spack/jobs/210304948) is a link to the error log, which seems about an invalid unicode to ascii conversion and unrelated to the PR. Also, the problem seems to have regressed on the current PRs. Honestly I can't understand from the log what can be the character that started the issue.

**Comment by adamjstewart** [comment]
If you look in the commit history, #3421 also shows the same Travis CI error.

**Comment by adamjstewart** [comment]
Checked out develop and ran the tests locally and they seem to pass. I think it's a false alarm.

**Comment by adamjstewart** [comment]
If anything it might mean that the changes in #3375 are non-deterministic.

**Comment by alalazo** [comment]
@adamjstewart I'll try to run the test under question repeatedly, and see if I can get an error locally.

**Comment by tgamblin** [comment]
I tried to run this locally and the tests succeed, but restarting any build in Travis fails.  Hm.  I will try harder.  Notable: both the mac and ubuntu travis builds fail.
